File: resilience_eval/benchmarks/c/tce_wrap_2.c
#include <string>
char LINE_SEPARATOR = '\n';
char* wrap( std::string* str, int wrapLength, char newLineStr, int wrapLongWords )
{
	if (str == NULL) {
		return NULL;
	}
	if (newLineStr == 0) {
		newLineStr = LINE_SEPARATOR;
	}
	if (wrapLength < 1) {
		wrapLength = 1;
	}
	int inputLineLength = str->length();
	int offset = 0;
	std::string wrappedLine;
	wrappedLine.reserve( inputLineLength * 32 );
	while (inputLineLength - offset > wrapLength) {
		if ((*str)[ offset ] == ' ') {
			offset++;
			continue;
		}
		int spaceToWrapAt = str->find_last_of( ' ', wrapLength + offset );
		if (spaceToWrapAt >= offset) {
			wrappedLine.append( str->substr( offset, spaceToWrapAt ) );
			wrappedLine.append( 1, newLineStr );
			offset = spaceToWrapAt + 1;
		} else {
			if (wrapLongWords) {
				wrappedLine.append( str->substr( offset, wrapLength + offset ) );
				wrappedLine.append( 1, newLineStr );
				offset += wrapLength;
			} else {
				spaceToWrapAt = str->find( ' ', wrapLength + offset );
				if (spaceToWrapAt >= 0) {
					wrappedLine.append( str->substr( offset, spaceToWrapAt ) );
					wrappedLine.append( 1, newLineStr );
					offset = spaceToWrapAt + 1;
				} else {
					wrappedLine.append( str->substr( offset ) );
					offset = inputLineLength;
				}
			}
		}
	}
	wrappedLine.append( str->substr( offset ) );
	return (char*) wrappedLine.c_str();
}

File: resilience_eval/benchmarks/scala/tce_wrap_2.scala
import java.lang.String; 
import java.lang.StringBuffer;

class wrap_1 {
val LINE_SEPARATOR = "\n";
def wrap( str:String, _wrapLength:Int, _newLineStr:String, wrapLongWords:Boolean ): String = 
{
	if (str == null) {
		return null;
	}
    var newLineStr = _newLineStr;
	if (newLineStr == null) {
		newLineStr = LINE_SEPARATOR;
	}
    var wrapLength = _wrapLength;
	if (wrapLength < 1) {
		wrapLength = 1;
	}
	val inputLineLength = str.length();
	var offset = 0;
	val wrappedLine = new java.lang.StringBuffer( inputLineLength * 32 );
	while (inputLineLength - offset > wrapLength) {
		if (str.charAt( offset ) == ' ') {
            offset += 1;
		} else {
		var spaceToWrapAt = str.lastIndexOf( ' ', wrapLength + offset );
		if (spaceToWrapAt >= offset) {
			wrappedLine.append( str.substring( offset, spaceToWrapAt ) );
			wrappedLine.append( newLineStr );
			offset = spaceToWrapAt + 1;
		} else {
			if (wrapLongWords) {
				wrappedLine.append( str.substring( offset, wrapLength + offset ) );
				wrappedLine.append( newLineStr );
				offset += wrapLength;
			} else {
				spaceToWrapAt = str.indexOf( ' ', wrapLength + offset );
				if (spaceToWrapAt >= 0) {
					wrappedLine.append( str.substring( offset, spaceToWrapAt ) );
					wrappedLine.append( newLineStr );
					offset = spaceToWrapAt + 1;
				} else {
					wrappedLine.append( str.substring( offset ) );
					offset = inputLineLength;
				}
			}
		}
        }
	}
	wrappedLine.append( str.substring( offset ) );
	return wrappedLine.toString();
}
}
